neptune/internal/storage/storage_utils.py

```diff
@@ -107,28 +110,35 @@ def scan_unique_upload_entries(upload_entries):
     """
     walked_entries = set()
     for entry in upload_entries:
-        if os.path.isdir(entry.source_path):
+        if entry.is_stream() or not os.path.isdir(entry.source_path):
+            walked_entries.add(entry)
+        else:
             for root, _, files in os.walk(entry.source_path):
                 path_relative_to_entry_source = os.path.relpath(root, entry.source_path)
                 target_root = os.path.normpath(os.path.join(entry.target_path, path_relative_to_entry_source))
                 for filename in files:
                     walked_entries.add(UploadEntry(os.path.join(root, filename), os.path.join(target_root, filename)))
-        else:
-            walked_entries.add(entry)
+
     return walked_entries
 
 
 def split_upload_files(upload_entries, max_package_size=1 * 1024 * 1024, max_files=500):
     current_package = UploadPackage()
 
     for entry in upload_entries:
-        size = os.path.getsize(entry.source_path)
-
-        if (size + current_package.size > max_package_size or current_package.len > max_files) \
-                and not current_package.is_empty():
+        if entry.is_stream():
             yield current_package
             current_package.reset()
-        current_package.update(entry, size)
+            current_package.update(entry, 0)
+            yield current_package
+            current_package.reset()
```

Comment:
Since user can provide an open file/stream, is it possible that resetting will fail?

On certain server response codes we repeat the upload, won't this invalidate the non-resettable steam and cause uploaded file to be corrupted?